cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -0,0 +1,699 @@
+    auto builder = static_cast<Date64Builder*>(abuilder);
+    auto batch = static_cast<liborc::LongVectorBatch*>(cbatch);
+
+    if (length == 0) {
+      return Status::OK();
+    }
+    int start = builder->length();
+
+    const uint8_t* valid_bytes = nullptr;
+    if (batch->hasNulls) {
+      valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
+    }
+    RETURN_NOT_OK(builder->AppendNulls(valid_bytes, length));
+
+    const int64_t* source = batch->data.data() + offset;
+    int64_t* target = reinterpret_cast<int64_t*>(builder->data()->mutable_data());
+
+    for (int i = 0; i < length; i++) {
+      target[start + i] = source[i] * 24 * 60 * 60 * 1000;
+    }
+    return Status::OK();
+  }
+
+  template <class builder_type>
+  Status append_binary_batch(ArrayBuilder* abuilder, liborc::ColumnVectorBatch* cbatch,
+                             int64_t offset, int64_t length) {
+    auto builder = static_cast<builder_type*>(abuilder);
+    auto batch = static_cast<liborc::StringVectorBatch*>(cbatch);
+
+    for (int i = offset; i < length + offset; i++) {
+      if (!batch->hasNulls || batch->notNull[i]) {
+        RETURN_NOT_OK(builder->Append(batch->data[i], batch->length[i]));
+      } else {
+        RETURN_NOT_OK(builder->AppendNull());
+      }
+    }
+    return Status::OK();
+  }
+
+  Status append_fixed_binary_batch(ArrayBuilder* abuilder,
+                                   liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                                   int64_t length) {
+    auto builder = static_cast<FixedSizeBinaryBuilder*>(abuilder);
+    auto batch = static_cast<liborc::StringVectorBatch*>(cbatch);
+
+    for (int i = offset; i < length + offset; i++) {
+      if (!batch->hasNulls || batch->notNull[i]) {
+        RETURN_NOT_OK(builder->Append(batch->data[i]));
+      } else {
+        RETURN_NOT_OK(builder->AppendNull());
+      }
+    }
+    return Status::OK();
+  }
+
+  Status append_decimal_batch(const liborc::Type* type, ArrayBuilder* abuilder,
+                              liborc::ColumnVectorBatch* cbatch, int64_t offset,
+                              int64_t length) {
+    auto builder = static_cast<Decimal128Builder*>(abuilder);
+
+    if (type->getPrecision() == 0 || type->getPrecision() > 18) {
+      auto batch = static_cast<liborc::Decimal128VectorBatch*>(cbatch);
+      for (int i = offset; i < length + offset; i++) {
+        if (!batch->hasNulls || batch->notNull[i]) {
+          const Decimal128& x =
+              Decimal128(batch->values[i].getHighBits(), batch->values[i].getLowBits());
+          RETURN_NOT_OK(builder->Append(x));
+        } else {
+          RETURN_NOT_OK(builder->AppendNull());
+        }
+      }
+    } else {
+      auto batch = static_cast<liborc::Decimal64VectorBatch*>(cbatch);
+      for (int i = offset; i < length + offset; i++) {
+        if (!batch->hasNulls || batch->notNull[i]) {
+          const Decimal128& x = Decimal128(batch->values[i]);
```

Comment:
Could this yield undefined behavior? Here either `x` should be properly on the stack or passed as an rvalue to `Append` (`Append(Decimal128(...))`)

Reply:
That's actually valid behavior per the standard, but I think it's more complicated than it needs to be. This should be a stack variable or temporary.